### ISSUE TITLE
data events can be fired after exit is fired

### ISSUE THREAD

**Issue by Tyriar** [microsoft/node-pty#72]
Downstream issue/repro: https://github.com/Microsoft/vscode/issues/23051
Temporary fix: https://github.com/Microsoft/vscode/commit/9464b54f39d8db943ddd4c134d9bef835b7bd506

Relevant commit that tried to fix the problem: https://github.com/chjj/pty.js/commit/4af628538ff5f970c2c84da31ba1e20db62696c0

**Comment by jerch** [comment]
@Tyriar Any chance to reproduce it with a simpler case? Imho it is related to the async data cycling (with a possible buffer delay), while the exit gets through earlier. Might be related to the whole socket thingy and open pandora's box ;)

**Comment by Tyriar** [comment]
I expect the repro would work when plugged into examples/fork/index, by setting shell to `"cmd.exe"` and args to `["-c", "ls -lR"]` then killing the process right when exit is fired.

**Comment by jerch** [comment]
@Tyriar Not sure yet if it is related, but if I run the following snippet under linux, the output will stop at different positions in data and kinda never finishes output:
```js
var os = require('os');
var pty = require('./lib/index');

var ptyProcess = pty.spawn("ls", ["-lR", "/usr/lib"], {
  name: 'xterm-color',
  cols: 80,
  rows: 30,
  cwd: process.env.HOME,
  env: process.env
});

ptyProcess.on('data', function(data) {
  console.log(data);
});

setTimeout(function(){}, 5000);
```

**Comment by jerch** [comment]
I drilled down the last problem to node's stream handling - neither `net.Socket` nor `tty.ReadStream` will reliably output the full data. The exit of the slave program kinda always happens prior all data got handled. It seems to be a race condition between node's non blocking stream handling and the closing of the pty pipe by the exit of the slave program. Gonna try to work around this by keeping the slave fd open until no further data can be read. Not sure yet if this will help to solve the original problem.

**Comment by jerch** [comment]
@Tyriar  What is the idea behind the 'exit' event of pty? Shall it mark the real process end (as it does atm more or less) or shall it be the last event of any possible interaction with pty (i.e. all pending data read)?

What happens under linux atm is somewhat unreliable due to the buffers and the async callbacks involved:
- slave program runs and generates output --> output gets catched by `net.Socket`, buffers the data and generates appropriate 'data' events
- program gets killed, the module does this atm by sending `SIGHUP` to the child via `process.kill(this.pid, 'SIGHUP')` --> gets invoked immediately by libuv
- program stops --> slave end of the pty device is closed, the async waitpid function set up by `PtyFork` gets invoked and cleans up the process data, sets the return status code *AND* installs the 'onexit' handler to the libuv loop to be invoked somewhere in the near future
- next step: *RACE CONDITION*
   - condition 1: the 'onexit' handler runs (does whatever was registered there) prior the next `net.Socket` data delivery invocation
       - subcondition: `net.Socket` internal buffers are empty, nomore data can be delivered since the pty device is closed - imho a wanted case
       - subcondition: `net.Socket` internal buffers are not empty and leftover data gets delivered - seems to be the issue case
   - condition 2: `net.Socket` data delivery has no pending data and stumbles over the already closed pty pipe - the socket will trigger an 'error' event with 'EIO', the 'onexit' handler gets invoked afterwards - seems to me to be a wanted case

Back to the question above - if the 'exit' event shall mark the stop of the slave program the current behavior is correct, the program exit will happen at any time while there still might be unhandled data. If the 'exit' shall be the final event it might be better to attach it the the 'end' of the pty ReadStream.

To be able to get all output data from slave until the exit happened, some more work is needed. This is due to the fact that any buffered data in pty device itself (not yet in `net.Socket`) will be lost as soon as any last endpoint gets closed.

**Comment by Tyriar** [comment]
Yes I think the pty's exit event should be triggered after the process is exited and all data events have been fired.

**Comment by alexxed** [comment]
2 years after, this is still an issue, and the one fix that worked for me is https://github.com/chjj/pty.js/pull/110#issuecomment-406470140

**Comment by jackyzha0** [comment]
we also ran into this problem in our own pty implementation https://github.com/replit/ruspty and fixed it via a 'synthetic eof' approach: https://github.com/replit/ruspty/pull/93

just dropping how we solved it here in case this is helpful

**Comment by jerch** [comment]
I think the issue here can be reduced to a specialty of linux pipes - POLLHUP does not necessarily mark the end of the data, but can happen earlier indicating the hang up of the other side indepedently. When this happens, the stream is marked as ended, but might still hold data in data buffers.

Imho libuv does not handle this POLLHUP edge case and will just send the exit event right away and the remaining data chunks afterwards.

Note that this is a linux specialty, on BSD posix POLLHUP will not be signalled until all data was drained leaving the logical order of events intact. (Tested that with macos and FreeBSD, so there is still a chance that other more exotic BSDs behave differently.)

**Comment by jackyzha0** [comment]
yes but you can kind of trick the kernel by keeping an extra reference to the user fd open so that when the program exits, the user fd is not closed immediately

normally what happens is that the child process would exit and close its inherited user fd which in kernel code immediately signals EIO which libuv will interpret as immediately the end of the stream even if there is data left in the pipe

by injecting a sentinel at the end of the data and having node be responsible for dropping the last reference to the user fd, we can guarantee that node reads everything in the pipe

**Comment by jerch** [comment]
@jackyzha0 Yes the second slave fd trick should do here, but how do you know when to drop the own fd? It smells a bit like a chance to leak an fd here?

Wouldn't it be easier to hold the exit signal back until all data was drained?

(On a sidenote - the second slave fd trick is the only way to properly maintain a PTY connection on solaris, as the termios `winsize` API only works on slaves there.)

**Comment by jackyzha0** [comment]
controller fd is not given to the child and we have the pty struct on the rust side hold ownership of the controller fd until the nodejs wrapper takes ownership and converts it to a node:tty ReadStream

we attach an error listener on the ReadStream specifically for EIO on the controller (this happens when the user fd is dropped, kernel sets TTY_OTHER_CLOSED and so next read on controller fd returns EIO) and we manually close the ReadStream which drops the controller fd

afaik theres no way to 'block' a program from exiting until _both_ fds are drained (O_NONBLOCK clear means at least all the data will be written to the user fd by the time program exits but no guarantees on whether that data has made it through to the controller fd and libuv)

to go back on your fd leaking question, we did worry about the user fd being held on forever (in the case for example of the sentinel write to the user fd failing due to repeated EINTR / write(2) returning 0) in which case we just have nodejs wait up to 1s on all the data before giving up and yanking the user fd 🤷

**Comment by jerch** [comment]
> afaik theres no way to 'block' a program from exiting until both fds are drained (O_NONBLOCK clear means at least all the data will be written to the user fd by the time program exits but no guarantees on whether that data has made it through to the controller fd and libuv)

Thats not what i meant, it is clear, that you cannot stop a program from exiting whenever it decides to do so (unless you have a debugging breakpoint on it).

The problem of this issue here is that the logical order of `exit` and `data` events is mixed up or at least unexpected due to the way linux treats POLLHUP independently. So I my suggestion would be to hold back the `exit` event until all data is really drained from the "dying" master fd. This "finally drained" state is a bit hard to spot, but if I remember right a `read(0)` (in C/C++) on the drained pipe produces an EIO or EBADF at this point. `read(0)` is important to not mess up the poll impl within libuv getting the left-over chunks cleanly.

**Comment by jackyzha0** [comment]
yes thats what we do in our implementation, we dont emit exit from the stream until we receive our synthetic eof

**Comment by jerch** [comment]
I still dont think that the second fd is needed, even more i think it makes this more unreliable, as POLLHUP would not occur anymore the same way as it would without that second fd. But if it works for you and you get everything properly cleaned up afterwards - then prolly stick with your solution.

**Comment by Tyriar** [comment]
This is how ruspty (stripped down version of node-pty without Windows support) handles this: https://github.com/replit/ruspty/pull/93
